### PR TITLE
TECHNICAL IMPROVEMENT: Updates to Rails 6 and Spree 4

### PR DESCRIPTION
**Pull request by robert-hromej**

### CODE REVIEW

**Comment by robert-hromej** [comment]
[Replace class_eval/module_eval for spree_avatax_certified](https://trello.com/c/LLXsb79k/1726-replace-classeval-moduleeval-for-spreeavataxcertified)

**Review by viktordanch — APPROVED** [review]

**Review by balazslaszlo — APPROVED** [review]

File: app/models/spree/address_decorator.rb [hunk 1]
```diff
@@ -1,13 +1,20 @@
...
+      Spree::Address.validation_enabled_countries.include?(country.try(:name))
+    end
+
+    def self.validation_enabled_countries
+      Spree::Config.avatax_address_validation_enabled_countries
+    end
   end
 end
+
+Spree::Address.prepend Spree::AddressDecorator
```
Comment (rczjns) [inline app/models/spree/address_decorator.rb]: add a newline to the end, please

File: app/models/spree/adjustment_decorator.rb [hunk 2]
```diff
@@ -1,10 +1,19 @@
...
 
-  def avatax_cache_key
-    key = ['Spree::Adjustment']
-    key << id
-    key << amount
-    key.join('-')
+module Spree
+  # AdjustmentDecorator
+  module AdjustmentDecorator
+    def self.prepended(base)
```
Comment (rczjns) [inline app/models/spree/adjustment_decorator.rb:5]: I'm thinking about is it possible to use `extend module` for a class to add a new class method for that?

e.g.
```
module XClassDecorator
  scope :not_tax, -> { where.not(source_type: 'Spree::TaxRate') }
end

X.extend XClassDecorator
```

I think it's more readable than `def self.prepended(base)`.
What do you think?

File: app/models/spree/order_decorator.rb [hunk 3]
```diff
@@ -1,74 +1,86 @@
...
-                                      :if => :address_validation_enabled?
-  def avalara_tax_enabled?
-    Spree::Config.avatax_tax_calculation
-  end
+# frozen_string_literal: true
+
+module Spree
+  # OrderDecorator
+  module OrderDecorator
+    def self.prepended(base)
```
Comment (rczjns) [inline app/models/spree/order_decorator.rb:5]: same like above

File: app/models/spree/order_decorator.rb [hunk 4]
```diff
@@ -1,74 +1,86 @@
...
 
-  def logger
-    @logger ||= SpreeAvataxCertified::AvataxLog.new('Spree::Order class', 'Start order processing')
+    def logger
+      @logger ||= SpreeAvataxCertified::AvataxLog.new('Spree::Order class', 'Start order processing')
+    end
   end
 end
+
+Spree::Order.prepend Spree::OrderDecorator
```
Comment (rczjns) [inline app/models/spree/order_decorator.rb]: new line please

File: app/models/spree/shipment_decorator.rb [hunk 5]
```diff
@@ -1,27 +1,35 @@
...
-    key << promo_total
-    key.join('-')
-  end
+# frozen_string_literal: true
 
-  def avatax_line_code
-    'FR'
-  end
+module Spree
+  # ShipmentDecorator
```
Comment (rczjns) [inline app/models/spree/shipment_decorator.rb]: I think it makes no sense to copy the name of the class/module as documentation...

**Review by rczjns — CHANGES_REQUESTED** [review]

File: app/models/spree/adjustment_decorator.rb (hunk 2 above)
Comment (robert-hromej) [reply app/models/spree/adjustment_decorator.rb:5]: @rczjns `extend module` - not work in this example.
```
$ irb
irb(main):001:0> class X
irb(main):002:1>   def self.scope(args)
irb(main):003:2>     puts "call scope method with args: '#{args}'"
irb(main):004:2>   end
irb(main):005:1> 
irb(main):006:1>   scope 'from X class'
irb(main):007:1> end
call scope method with args: 'from X class'
=> nil
irb(main):008:0> 
irb(main):009:0> p X.new.respond_to?(:instance_method_a)
false
=> false
irb(main):010:0> p X.respond_to?(:klass_method)
false
=> false
irb(main):011:0> module XClassDecorator
irb(main):012:1>   module ClassMethods
irb(main):013:2>     def klass_method(args)
irb(main):014:3>       # ...
irb(main):015:3>     end
irb(main):016:2>   end
irb(main):017:1> 
irb(main):018:1>   def self.prepended(base)
irb(main):019:2>     base.scope 'from XClassDecorator'
irb(main):020:2>     base.singleton_class.prepend(ClassMethods)
irb(main):021:2>   end
irb(main):022:1> 
irb(main):023:1>   def instance_method_a(args)
irb(main):024:2>     # ...
irb(main):025:2>   end
irb(main):026:1> end
=> :instance_method_a
irb(main):027:0> X.prepend XClassDecorator
call scope method with args: 'from XClassDecorator'
=> X
irb(main):028:0> p X.new.respond_to?(:instance_method_a)
true
=> true
irb(main):029:0> p X.respond_to?(:klass_method)
true
=> true
irb(main):030:0> 
```

I don't like this style(`prepended`, `ClassMethods`....) but only this work correctly

Comment (robert-hromej) [reply app/models/spree/adjustment_decorator.rb:5]: @rczjns I tried with extend module, but not working

**Review by rczjns — APPROVED** [review]

**Comment by robert-hromej** [comment]
[Migrate to Zeitwerk loader](https://trello.com/c/E8pZ2kZS/1723-migrate-to-zeitwerk-loader)

**Review by padavanocskos — APPROVED** [review]

**Review by balazslaszlo — APPROVED** [review]

**Pull request merged**